subsys/bluetooth/host/adv.c

```diff
@@ -462,20 +492,29 @@ static int hci_set_adv_ext_fragmented(struct bt_le_ext_adv *adv, uint16_t hci_op
 
 			/* We can't necessarily set one AD field in a single step. */
 			while (offset < data[j].data_len) {
-				buf = bt_hci_cmd_create(hci_op, sizeof(*set_data));
+				const size_t cmd_size =
+					MAX(sizeof(struct bt_hci_cp_le_set_per_adv_data),
+					    MAX(sizeof(struct bt_hci_cp_le_set_ext_adv_data),
+						sizeof(struct bt_hci_cp_le_set_ext_scan_rsp_data)));
+
+				buf = bt_hci_cmd_create(hci_op, cmd_size);
```

Comment:
Is the code somewhere overwriting the length field that's part of the HCI command header? The `bt_hci_cmd_create()` function encodes its second parameter into that header field, however at this point it seems the code doesn't yet know what the final length of the command.

Reply:
Yes, it does. It sets the length to the maximum value (as done for other commands)

Reply:
I was mainly concerned with the `bt_hci_cp_le_set_per_adv_data` struct, since it's smaller than the two others. However, now that I checked the HCI spec it seems like there might be an off-by-one issue with the data field of this struct (defined as `uint8_t data[251]`. From the HCI spec:
`0 to 252 | The number of octets in the Advertising Data parameter`

If that were to be fixed then all the structs would have the same size.

Reply:
@jhedberg , that is already fixed as part of this PR :). Now the struct is defined to have to correct maximum data length of 252 bytes.

Reply:
Ok. I guess I was confused by this construction of MAX statements, since that seems to imply that they are not identical in size.

Reply:
Anyway, I suppose this entire concern/confusion will go away when the suggested refactoring gets done